src/Microsoft.Health.ControlPlane.CosmosDb/Features/Storage/ControlPlaneDataStore.cs

```diff
@@ -0,0 +1,142 @@
+
+            var cosmosIdentityProvider = new CosmosIdentityProvider(identityProvider);
+            var resultIdentityProvider = await UpsertSystemObjectAsync(cosmosIdentityProvider, CosmosIdentityProvider.IdentityProviderPartition, cancellationToken);
+            return resultIdentityProvider.ToIdentityProvider();
+        }
+
+        internal IDocumentQuery<T> CreateDocumentQuery<T>(
+            SqlQuerySpec sqlQuerySpec,
+            FeedOptions feedOptions = null)
+        {
+            EnsureArg.IsNotNull(sqlQuerySpec, nameof(sqlQuerySpec));
+
+            var context = new CosmosQueryContext(_collectionUri, sqlQuerySpec, feedOptions);
+
+            return _cosmosDocumentQueryFactory.Create<T>(_documentClient.Value, context);
+        }
+
+        private async Task<T> GetSystemDocumentByIdAsync<T>(string id, string partitionKey, CancellationToken cancellationToken)
+        {
+            EnsureArg.IsNotNull(id, nameof(id));
+            EnsureArg.IsNotNull(partitionKey, nameof(partitionKey));
+            var documentQuery = new SqlQuerySpec(
+                "SELECT * FROM root r WHERE r.id = @id",
+                new SqlParameterCollection(new[] { new SqlParameter("@id", id) }));
+            var feedOptions = new FeedOptions
+            {
+                PartitionKey = new PartitionKey(partitionKey),
+            };
+            IDocumentQuery<T> cosmosDocumentQuery =
+                CreateDocumentQuery<T>(documentQuery, feedOptions);
+            using (cosmosDocumentQuery)
+            {
+                var response = await cosmosDocumentQuery.ExecuteNextAsync(cancellationToken);
+                var result = response.SingleOrDefault();
+                if (result == null)
+                {
+                    _logger.LogError($"{typeof(T)} with id {id} was not found in Cosmos DB.");
+                }
+
+                return result;
+            }
+        }
+
+        private async Task<T> UpsertSystemObjectAsync<T>(T systemObject, string partitionKey, CancellationToken cancellationToken)
+            where T : class
+        {
+            EnsureArg.IsNotNull(systemObject, nameof(systemObject));
+            var eTagAccessCondition = new AccessCondition();
+
+            var requestOptions = new RequestOptions
+            {
+                PartitionKey = new PartitionKey(partitionKey),
+                AccessCondition = eTagAccessCondition,
+            };
+            try
+            {
+                var response = await _documentClient.Value.UpsertDocumentAsync(
+                    _collectionUri,
+                    systemObject,
+                    requestOptions,
+                    true,
+                    cancellationToken);
+                _logger.LogInformation("Request charge: {RequestCharge}, latency: {RequestLatency}", response.RequestCharge, response.RequestLatency);
+                return (dynamic)response.Resource;
```

Comment:
nit: (T)(dynamic)response.Resource so that if it fails to cast for whatever reason it fails fast rather than somwhere down the line when property is accessed.